### PR TITLE
pam: trigger pam_authenticate on login

### LINKED ISSUE #3929: Support External PAM Auth Modules
### Feature Request
We've a customer who would like to enable a 3rd party Auth PAM module, specifically https://duo.com/docs/duounix#overview

### Motivation
By adding another check prior to a session starts, Teleport users are able to have another 2nd factor check. We can start by integrating with another provider, and should consider brining it in-house as part of https://github.com/gravitational/teleport/issues/3878 

### Who's it for?
OSS User, Pro, Enterprise

### PR DESCRIPTION
**Pull request by awly**
This will trigger any "auth" PAM modules configured on the system for
teleport. For example, Duo 2FA prompt on each connection.
The module will be able to interact with the user (e.g. print prompts).

Also, make PAM env var propagation consistent for port forwarding
sessions.

Fixes https://github.com/gravitational/teleport/issues/3929

### CODE REVIEW

**Comment by awly** [comment]
Tested with basic `pam_exec.so` module and [Duo](https://duo.com/docs/duounix#overview)

**Comment by russjones** [comment]
Can you add test coverage, we have a PAM module we use for testing that can be configured to test if Teleport calls the authentication function. https://github.com/gravitational/teleport/blob/master/modules/pam_teleport/pam_teleport.c

**Comment by webvictim** [comment]
Some weird test stuff happening here:

Jenkins:
```
signal: killed
FAIL	github.com/gravitational/teleport/lib/pam	74.098s
```

It looks like Drone timed out and killed the container too.

**Comment by webvictim** [comment]
Looks like a memory leak:

2020-07-02-141610__2247x495__maim

**Comment by awly** [comment]
> Looks like a memory leak:

I think it's because we're reading from `stdin` in some cases and in tests it's set to [an infinite nop reader](https://github.com/gravitational/teleport/blob/master/lib/pam/pam_test.go#L118-L123).

Added an `io.LimitReader` around it for sanity.

@russjones refactored the testing setup and added missing test cases.

**Comment by awly** [comment]
Need to re-run `make install` in `modules/pam_teleport` on the build box to get things passing again

**Comment by awly** [comment]
ok, I fixed the buildbox build steps, but looks like drone uses a cached version of buildbox.
@webvictim any particular reason we don't use `make -C build.assets test` here: https://github.com/gravitational/teleport/blob/master/.drone.yml#L77-L85 ?
That would let drone pick up any new changes to buildbox.

**Comment by webvictim** [comment]
The reason is mostly because Jenkins uses `/gopath` while Drone uses `/go`, so the Docker `-v` commands are different. We also use a shared Go build cache in Drone for speed which Jenkins doesn’t do. I can always modify the `Makefile` to hold the logic and just call those targets from Drone - the reason I didn’t initially is that for rapid iteration, it can be nice to be able to change the build code without having to cut new tags and backport fixes to release branches.

With that said, this has actually got me wondering whether the shared build cache could be influencing the test failures in Drone...

**Comment by awly** [comment]
Looks like https://github.com/gravitational/teleport/pull/3982 will soon fix the problem.
I'll wait for it to land, thanks @webvictim!

File: build.assets/Makefile
Comment (webvictim) [inline build.assets/Makefile:229]: As this is the Makefile in `build.assets`, the PAM build commands here should arguably be run inside a Docker container to reduce the need to have `pam-devel` or similar packages installed on the host.

File: lib/pam/pam_test.go
Comment (webvictim) [inline lib/pam/pam_test.go:101]: Are we now assuming that every single binary is built with PAM support, so tests will fail if it's not present? Or is this controlled by the `pam` tag?

Comment (awly) [reply lib/pam/pam_test.go:101]: This check is performed in `SetUpSuite`.
If the skip happens there, no tests will run.
There's no need to do the same check in every test now.

File: build.assets/Makefile
Comment (awly) [reply build.assets/Makefile]: Great idea!
Moved `modules/pam_teleport` to `build.assets/pam` and now it'll compile `pam_teleport.so` as part of buildbox creation.

**Comment by benarent** [comment]
Do Teleport users need to set anything, or is this automatically on with

```
  pam:
     enabled: true
```

**Comment by awly** [comment]
@benarent automatically on with that config field, yes.
Users do need to configure a PAM auth policy to get any effects from this.

**Comment by webvictim** [comment]
@awly #3982 is merged; I fixed the Dockerfile merge conflicts for you.

Looks like integration tests are failing.

**Comment by benarent** [comment]
👍 from me. I tested it with Duo and it worked perfectly. New Video Demo: https://share.getcloudapp.com/YEupwAAl

File: build.assets/Dockerfile
Comment (webvictim) [inline build.assets/Dockerfile:50]: This should be `VOLUME ["/go/src/github.com/gravitational/teleport"]` rather than `VOLUME ["/gopath/src/github.com/gravitational/teleport"]` - might be an idea to rebase?

Comment (awly) [reply build.assets/Dockerfile]: Done

**Review by webvictim — APPROVED** [review]

**Comment by awly** [comment]
Need approval from @klizhentas or @russjones on this one

**Review by russjones — APPROVED** [review]

**Pull request merged**
